File: truss-compiler/src/main/java/truss/compiler/symbols/NamespaceSymbol.java
package truss.compiler.symbols;

import org.apache.commons.lang.Validate;
import truss.compiler.Message;
import truss.compiler.MessageCollectionScope;
import truss.compiler.MessageType;
import truss.compiler.syntax.IdentifierNameSyntax;
import truss.compiler.syntax.NameSyntax;
import truss.compiler.syntax.NamespaceDeclarationSyntax;
import truss.compiler.syntax.QualifiedNameSyntax;

public class NamespaceSymbol extends ContainerSymbol {
    private final String name;

    protected NamespaceSymbol(String name, ContainerSymbol container) {
        super(container);

        this.name = name;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public SymbolKind getKind() {
        return SymbolKind.NAMESPACE;
    }

    @Override
    public void accept(SymbolVisitor visitor) {
        if (!visitor.isDone()) {
            visitor.visitNamespace(this);
        }
    }

    public static NamespaceSymbol fromDeclaration(NamespaceDeclarationSyntax syntax, ContainerSymbol container) {
        Validate.notNull(syntax, "syntax");
        Validate.notNull(container, "container");

        NamespaceSymbol symbol = createChain(syntax.getName(), container);

        symbol.addSpan(syntax.getSpan());

        return symbol;
    }

    private static NamespaceSymbol createChain(NameSyntax name, ContainerSymbol container) {
        if (name instanceof QualifiedNameSyntax) {
            QualifiedNameSyntax qualifiedName = (QualifiedNameSyntax)name;
            container = createChain(qualifiedName.getLeft(), container);
            name = qualifiedName.getRight();
        }

        if (!(name instanceof IdentifierNameSyntax)) {
            MessageCollectionScope.addMessage(new Message(
                MessageType.INVALID_NAMESPACE_IDENTIFIER,
                name.getSpan()
            ));

            return null;
        }

        String identifier = ((IdentifierNameSyntax)name).getIdentifier();

        for (Symbol member : container.getMemberByMetadataName(identifier)) {
            if (member instanceof NamespaceSymbol) {
                return (NamespaceSymbol)member;
            }
        }

        NamespaceSymbol symbol = new NamespaceSymbol(
            identifier,
            container
        );

        container.addMember(symbol);

        return symbol;
    }
}

File: Truss.Compiler/Symbols/NamespaceSymbol.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Symbols {
    public class NamespaceSymbol : ContainerSymbol {
        private readonly string _name;

        protected NamespaceSymbol(string name, ContainerSymbol container)
            : base(container) {
            _name = name;
        }

        public override string Name {
            get { return _name; }
        }

        public override SymbolKind Kind {
            get { return SymbolKind.Namespace; }
        }

        public override void Accept(ISymbolVisitor visitor) {
            if (!visitor.Done) {
                visitor.VisitNamespace(this);
            }
        }

        public static NamespaceSymbol FromDeclaration(ErrorList errors, NamespaceDeclarationSyntax syntax, ContainerSymbol container) {
            if (errors == null) {
                throw new ArgumentNullException("errors");
            }
            if (syntax == null) {
                throw new ArgumentNullException("syntax");
            }
            if (container == null) {
                throw new ArgumentNullException("container");
            }

            return CreateChain(errors, syntax.Name, container);
        }

        private static NamespaceSymbol CreateChain(ErrorList errors, NameSyntax name, ContainerSymbol container) {
            if (name is QualifiedNameSyntax) {
                var qualifiedName = (QualifiedNameSyntax)name;
                container = CreateChain(errors, qualifiedName.Left, container);
                name = qualifiedName.Right;
            }

            if (!(name is IdentifierNameSyntax)) {
                errors.Add(Error.InvalidNamespaceIdentifier, name.Span);

                return null;
            }

            string identifier = ((IdentifierNameSyntax)name).Identifier;

            NamespaceSymbol symbol;
            if (!container.TryGetMemberByMetadataName(identifier, out symbol)) {
                symbol = new NamespaceSymbol(identifier, container);

                container.AddMember(symbol);
            }

            return symbol;
        }
    }
}
